### PR TITLE
feature[cartesian]: Add support for HIP/ROCm

### PR DESCRIPTION
**Pull request by stubbiali**
Adaptations to generate GridTools/DaCe code targeting AMD GPUs. Complemented by https://github.com/GridTools/gridtools/pull/1759.

### CODE REVIEW

File: src/gt4py/cartesian/config.py [hunk 1]
```diff
@@ -32,6 +32,8 @@
 
 CUDA_HOST_CXX: Optional[str] = os.environ.get("CUDA_HOST_CXX", None)
 
+GT4PY_USE_HIP: int = int(os.environ.get("GT4PY_USE_HIP", 0))
```
Comment (egparedes) [inline src/gt4py/cartesian/config.py:35]: There should be a way to autodetect if cupy has been compiled with HIP. This is a naive suggestion to do it using dlpack, but probably there is a simpler way using some esoteric cupy API endpoint.
```suggestion
if "GT4PY_USE_HIP" in os.environ:
    GT4PY_USE_HIP: bool = bool(int(os.environ["GT4PY_USE_HIP"])) 
else:
    # Autodetect cupy with ROCm/HIP using DLPack device type
    # Reference: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack_device__.html#array_api.array.__dlpack_device__
    try:
        import cupy as cp
        GT4PY_USE_HIP = cp.empty((1,)).__dlpack_device__()[0] == 10
    except Exception:
        GT4PY_USE_HIP = False
```

**Review by egparedes — COMMENTED** [review]
Not a review, just a couple of suggestions

File: src/gt4py/storage/utils.py [hunk 2]
```diff
@@ -243,6 +245,18 @@ def allocate_gpu(
     if device_field.ndim > 0:
         device_field = device_field[tuple(slice(0, s, None) for s in shape)]
 
+    if gt_config.GT4PY_USE_HIP:
+        # HIP/ROCm lack support for __cuda_array_interface__
+        device_field.__hip_array_interface__ = {
```
[the remark this answers is not in the document]
Comment (gronerl) [reply src/gt4py/storage/utils.py:250]: I agree, unless things have changed, it seems that the community just always uses the __cuda_array_interface__ name even for AMD device buffers. (e.g. most importantly cupy)

File: src/gt4py/cartesian/backend/pyext_builder.py [hunk 3]
```diff
@@ -85,43 +85,56 @@ def get_gt_pyext_build_opts(
...
+            extra_compile_args["cuda"].append(
+                "-isystem{}".format(os.path.join(dace_path, "runtime/include"))
+            )
+        else:
+            extra_compile_args["cxx"].append(
+                "-isystem{}".format(os.path.join(dace_path, "runtime/include"))
+            )
+            extra_compile_args["cuda"].append(
+                "-isystem{}".format(os.path.join(dace_path, "runtime/include"))
+            )
```
Comment (gronerl) [inline src/gt4py/cartesian/backend/pyext_builder.py:132]: these two branches are the same

**Review by gronerl — CHANGES_REQUESTED** [review]
Only a small comment. 
I wonder if there is a way at CSCS to run something on AMD as part of CI.

File: src/gt4py/storage/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gronerl) [reply src/gt4py/storage/utils.py:250]: My understanding of cupy/cupy#7431 is that this is some particular case where you would want to wrap cpp side allocated data in a cupy buffer. Have you tried and ran into the error? Last time we ran AMD a while back we used `__cuda_array_interface__` happily to pass cupy arrays to stencils.

[the remark this answers is not in the document]
Comment (stubbiali) [reply src/gt4py/storage/utils.py:250]: Yes, I tried and the GridTools bindings threw the exception `AttributeError: HIP/ROCm does not support cuda array interface`. Same happens if you try to read `__cuda_array_interface__` from within Python.

**Comment by FlorianDeconinck** [comment]
If this needs an extra set of tests NASA/NOAA can wiggle some AMD systems, hit me up.

(Great to see this!)

File: src/gt4py/storage/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (FlorianDeconinck) [reply src/gt4py/storage/utils.py:250]: +1 for DLPack.
I've seen accelerated adoption lately. Is it on the radar @egparedes  to add it ? 
Feels not urgent, but it's something we could contribute later down the road

[the remark this answers is not in the document]
Comment (gronerl) [reply src/gt4py/storage/utils.py:250]: If this error is there, is it safe to just work around it in this way? There must be a reason why it is not supported.

File: src/gt4py/cartesian/config.py (hunk 1 above)
Comment (stubbiali) [reply src/gt4py/cartesian/config.py]: I partially followed your suggestion using `cupy.cuda.get_hipcc_path()`.

File: src/gt4py/storage/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/storage/utils.py:250]: > +1 for DLPack. I've seen accelerated adoption lately. Is it on the radar @egparedes to add it ? Feels not urgent, but it's something we could contribute later down the road

Yes, we're looking into supporting DLPack. There is no timeline yet, but hopefully not far in the future.

**Comment by havogt** [comment]
cscs-ci run

File: src/gt4py/cartesian/backend/pyext_builder.py [hunk 4]
```diff
@@ -85,43 +85,48 @@ def get_gt_pyext_build_opts(
...
-        nvcc=[
-            "-std=c++17",
-            "-ftemplate-depth={}".format(gt_config.build_settings["cpp_template_depth"]),
+        ]
+    )
+    extra_compile_args["cuda"] = [
+        "-std=c++17",
+        "-ftemplate-depth={}".format(gt_config.build_settings["cpp_template_depth"]),
+        "-isystem{}".format(gt_include_path),
+        "-isystem{}".format(gt_config.build_settings["boost_include_path"]),
```
Comment (havogt) [inline src/gt4py/cartesian/backend/pyext_builder.py:94]:
```suggestion
        "-isystem={}".format(gt_include_path),
        "-isystem={}".format(gt_config.build_settings["boost_include_path"]),
```

File: src/gt4py/cartesian/backend/pyext_builder.py [hunk 5]
```diff
@@ -85,43 +85,48 @@ def get_gt_pyext_build_opts(
...
     extra_link_args.extend(mode_flags)
 
     if dace_path := get_dace_module_path():
         extra_compile_args["cxx"].append(
             "-isystem{}".format(os.path.join(dace_path, "runtime/include"))
         )
-        extra_compile_args["nvcc"].append(
-            "-isystem={}".format(os.path.join(dace_path, "runtime/include"))
+        extra_compile_args["cuda"].append(
+            "-isystem{}".format(os.path.join(dace_path, "runtime/include"))
```
Comment (havogt) [inline src/gt4py/cartesian/backend/pyext_builder.py:123]:
```suggestion
            "-isystem={}".format(os.path.join(dace_path, "runtime/include"))
```

**Review by havogt — CHANGES_REQUESTED** [review]
nvcc needs `-isystem=`

**Comment by havogt** [comment]
cscs-ci run

**Comment by havogt** [comment]
cscs-ci run

**Review by havogt — APPROVED** [review]

**Pull request merged**